File: server-csharp/Collisions/GJK/Methods.cs
using System.Diagnostics.Contracts;
using System.Numerics;
using SpacetimeDB;

public static partial class Module
{
    public static bool SolveGjk(List<ConvexHullCollider> ColliderA, DbVector3 PositionA, float YawRadiansA, List<ConvexHullCollider> ColliderB, DbVector3 PositionB, float YawRadiansB, out GjkResult Result, int MaxIterations = 32)
    {
        List<GjkVertex> Simplex = new List<GjkVertex>(4);
        DbVector3 SearchDirection = new DbVector3(0f, 0f, 1f);

        GjkVertex InitialVertex = SupportPairWorld(ColliderA, PositionA, YawRadiansA, ColliderB, PositionB, YawRadiansB, SearchDirection);
        float InitialDot = Dot(InitialVertex.MinkowskiPoint, SearchDirection);

        if (InitialDot <= 0f)
        {
            DbVector3 ClosestPointA;
            DbVector3 ClosestPointB;

            ComputeWitnessPointsFromSimplex(Simplex, out ClosestPointA, out ClosestPointB);

            Result = new GjkResult
            {
                Intersects = false,
                Simplex = Simplex,
                LastDirection = Negate(SearchDirection),
                ClosestPointA = ClosestPointA,
                ClosestPointB = ClosestPointB
            };
            return false;
        }

        Simplex.Add(InitialVertex);
        SearchDirection = Negate(InitialVertex.MinkowskiPoint);

        for (int IterationIndex = 0; IterationIndex < MaxIterations; IterationIndex++)
        {
            GjkVertex SupportVertex = SupportPairWorld(ColliderA, PositionA, YawRadiansA, ColliderB, PositionB, YawRadiansB, SearchDirection);
            float SupportDot = Dot(SupportVertex.MinkowskiPoint, SearchDirection);

            if (SupportDot <= 0f)
            {
                DbVector3 ClosestPointA;
                DbVector3 ClosestPointB;

                ComputeWitnessPointsFromSimplex(Simplex, out ClosestPointA, out ClosestPointB);

                Result = new GjkResult
                {
                    Intersects = false,
                    Simplex = Simplex,
                    LastDirection = Negate(SearchDirection),
                    ClosestPointA = ClosestPointA,
                    ClosestPointB = ClosestPointB
                };
                return false;
            }

            Simplex.Add(SupportVertex);

            if (UpdateSimplex(ref Simplex, ref SearchDirection))
            {
                DbVector3 ClosestPointA;
                DbVector3 ClosestPointB;

                ComputeWitnessPointsFromSimplex(Simplex, out ClosestPointA, out ClosestPointB);

                Result = new GjkResult
                {
                    Intersects = true,
                    Simplex = Simplex,
                    LastDirection = Negate(SearchDirection),
                    ClosestPointA = ClosestPointA,
                    ClosestPointB = ClosestPointB
                };
                return true;
            }
        }

        {
            DbVector3 ClosestPointA;
            DbVector3 ClosestPointB;

            ComputeWitnessPointsFromSimplex(Simplex, out ClosestPointA, out ClosestPointB);

            Result = new GjkResult
            {
                Intersects = false,
                Simplex = Simplex,
                LastDirection = Negate(SearchDirection),
                ClosestPointA = ClosestPointA,
                ClosestPointB = ClosestPointB
            };
            return false;
        }
    }

    static void ComputeWitnessPointsFromSimplex(List<GjkVertex> Simplex, out DbVector3 ClosestPointA, out DbVector3 ClosestPointB)
    {
        int Count = Simplex.Count;

        if (Count == 0)
        {
            ClosestPointA = new DbVector3(0f, 0f, 0f);
            ClosestPointB = new DbVector3(0f, 0f, 0f);
            return;
        }

        if (Count == 1)
        {
            GjkVertex V = Simplex[0];
            ClosestPointA = V.SupportPointA;
            ClosestPointB = V.SupportPointB;
            return;
        }

        int BestIndexI = 0;
        int BestIndexJ = 1;
        float BestDistanceSquared = float.PositiveInfinity;

        for (int I = 0; I < Count; I++)
        {
            DbVector3 Wi = Simplex[I].MinkowskiPoint;
            float DistanceSquaredI = Dot(Wi, Wi);

            if (DistanceSquaredI < BestDistanceSquared)
            {
                BestDistanceSquared = DistanceSquaredI;
                BestIndexI = I;
                BestIndexJ = I;
            }

            for (int J = I + 1; J < Count; J++)
            {
                DbVector3 W0 = Simplex[I].MinkowskiPoint;
                DbVector3 W1 = Simplex[J].MinkowskiPoint;

                DbVector3 Edge = Sub(W1, W0);
                float EdgeLengthSquared = Dot(Edge, Edge);

                if (EdgeLengthSquared <= 1e-12f)
                {
                    continue;
                }

                DbVector3 NegativeW0 = Negate(W0);
                float T = Dot(NegativeW0, Edge) / EdgeLengthSquared;

                if (T < 0f)
                {
                    T = 0f;
                }
                else if (T > 1f)
                {
                    T = 1f;
                }

                DbVector3 ClosestMinkowski = new DbVector3(
                    W0.x + Edge.x * T,
                    W0.y + Edge.y * T,
                    W0.z + Edge.z * T
                );

                float DistanceSquared = Dot(ClosestMinkowski, ClosestMinkowski);

                if (DistanceSquared < BestDistanceSquared)
                {
                    BestDistanceSquared = DistanceSquared;
                    BestIndexI = I;
                    BestIndexJ = J;
                }
            }
        }

        if (BestIndexI == BestIndexJ)
        {
            GjkVertex V = Simplex[BestIndexI];
            ClosestPointA = V.SupportPointA;
            ClosestPointB = V.SupportPointB;
        }
        else
        {
            GjkVertex V0 = Simplex[BestIndexI];
            GjkVertex V1 = Simplex[BestIndexJ];

            DbVector3 W0 = V0.MinkowskiPoint;
            DbVector3 W1 = V1.MinkowskiPoint;
            DbVector3 Edge = Sub(W1, W0);

            float EdgeLengthSquared = Dot(Edge, Edge);

            if (EdgeLengthSquared <= 1e-12f)
            {
                ClosestPointA = V0.SupportPointA;
                ClosestPointB = V0.SupportPointB;
                return;
            }

            DbVector3 NegativeW0 = Negate(W0);
            float T = Dot(NegativeW0, Edge) / EdgeLengthSquared;

            if (T < 0f)
            {
                T = 0f;
            }
            else if (T > 1f)
            {
                T = 1f;
            }

            DbVector3 SupportA0 = V0.SupportPointA;
            DbVector3 SupportA1 = V1.SupportPointA;
            DbVector3 SupportB0 = V0.SupportPointB;
            DbVector3 SupportB1 = V1.SupportPointB;

            ClosestPointA = new DbVector3(
                SupportA0.x + (SupportA1.x - SupportA0.x) * T,
                SupportA0.y + (SupportA1.y - SupportA0.y) * T,
                SupportA0.z + (SupportA1.z - SupportA0.z) * T
            );

            ClosestPointB = new DbVector3(
                SupportB0.x + (SupportB1.x - SupportB0.x) * T,
                SupportB0.y + (SupportB1.y - SupportB0.y) * T,
                SupportB0.z + (SupportB1.z - SupportB0.z) * T
            );
        }
    }

    static GjkVertex SupportPairWorld(List<ConvexHullCollider> ComplexColliderA, DbVector3 PositionA, float YawRadiansA, List<ConvexHullCollider> ComplexColliderB, DbVector3 PositionB, float YawRadiansB, DbVector3 DirectionWorld)
    {
        DbVector3 SupportPointAWorld = SupportWorldComplex(ComplexColliderA, PositionA, YawRadiansA, DirectionWorld);
        DbVector3 SupportPointBWorld = SupportWorldComplex(ComplexColliderB, PositionB, YawRadiansB, Negate(DirectionWorld));
        return new GjkVertex(SupportPointAWorld, SupportPointBWorld);
    }

    static DbVector3 SupportWorldComplex(List<ConvexHullCollider> ComplexCollider, DbVector3 WorldPosition, float YawRadians, DbVector3 DirectionWorld)
    {
        DbVector3 DirectionLocal = RotateAroundYAxis(DirectionWorld, -YawRadians);
        DbVector3 SupportLocalPoint = SupportLocalComplex(ComplexCollider, DirectionLocal);
        DbVector3 SupportWorldRotated = RotateAroundYAxis(SupportLocalPoint, YawRadians);
        return Add(SupportWorldRotated, WorldPosition);
    }

    static DbVector3 SupportLocalComplex(List<ConvexHullCollider> ComplexCollider, DbVector3 DirectionLocal)
    {
        float BestDot = float.NegativeInfinity;
        DbVector3 BestPoint = new(0f, 0f, 0f);

        for (int Index = 0; Index < ComplexCollider.Count; Index++)
        {
            DbVector3 HullSupportPoint = SupportLocal(ComplexCollider[Index], DirectionLocal);
            float DotValue = Dot(HullSupportPoint, DirectionLocal);

            if (DotValue > BestDot)
            {
                BestDot = DotValue;
                BestPoint = HullSupportPoint;
            }
        }

        return BestPoint;
    }

    static DbVector3 SupportLocal(ConvexHullCollider Collider, DbVector3 Direction)
    {
        List<DbVector3> Vertices = Collider.VerticesLocal;

        int BestVertexIndex = 0;
        float BestDotProduct = Dot(Vertices[0], Direction);

        for (int VertexIndex = 1; VertexIndex < Vertices.Count; VertexIndex++)
        {
            float DotProduct = Dot(Vertices[VertexIndex], Direction);
            if (DotProduct > BestDotProduct)
            {
                BestDotProduct = DotProduct;
                BestVertexIndex = VertexIndex;
            }
        }

        DbVector3 BestVertex = Vertices[BestVertexIndex];

        // Margin inflation (logical skin)
        float Margin = Collider.Margin;
        if (Margin > 0f)
        {
            float DirLenSq = Dot(Direction, Direction);
            if (DirLenSq > 1e-8f)
            {
                float InvLen = 1.0f / MathF.Sqrt(DirLenSq);
                DbVector3 DirNorm = new(Direction.x * InvLen, Direction.y * InvLen, Direction.z * InvLen);

                BestVertex = new DbVector3(
                    BestVertex.x + DirNorm.x * Margin,
                    BestVertex.y + DirNorm.y * Margin,
                    BestVertex.z + DirNorm.z * Margin
                );
            }
        }

        return BestVertex;
    }


    static DbVector3 RotateAroundYAxis(DbVector3 Vector, float YawRadians)
    {
        float CosYaw = MathF.Cos(YawRadians);
        float SinYaw = MathF.Sin(YawRadians);

        float RotatedX = Vector.x * CosYaw + Vector.z * SinYaw;
        float RotatedZ = -Vector.x * SinYaw + Vector.z * CosYaw;

        return new DbVector3(RotatedX, Vector.y, RotatedZ);
    }

    static bool UpdateSimplex(ref List<GjkVertex> Simplex, ref DbVector3 SearchDirection)
    {
        int SimplexCount = Simplex.Count;

        if (SimplexCount == 2)
        {
            DbVector3 PointA = Simplex[1].MinkowskiPoint;
            DbVector3 PointB = Simplex[0].MinkowskiPoint;

            DbVector3 SegmentBA = Sub(PointB, PointA);
            DbVector3 VectorToOriginFromA = Negate(PointA);

            float DotSegmentWithAO = Dot(SegmentBA, VectorToOriginFromA);

            if (DotSegmentWithAO <= 0f)
            {
                Simplex.RemoveAt(0);
                SearchDirection = VectorToOriginFromA;
                return false;
            }

            DbVector3 NewDirection = TripleCross(SegmentBA, VectorToOriginFromA, SegmentBA);

            if (NearZero(NewDirection))
            {
                NewDirection = Perp(SegmentBA);
            }

            SearchDirection = NewDirection;
            return false;
        }

        if (SimplexCount == 3)
        {
            DbVector3 PointA = Simplex[2].MinkowskiPoint;
            DbVector3 PointB = Simplex[1].MinkowskiPoint;
            DbVector3 PointC = Simplex[0].MinkowskiPoint;

            DbVector3 EdgeAB = Sub(PointB, PointA);
            DbVector3 EdgeAC = Sub(PointC, PointA);
            DbVector3 VectorToOriginFromA = Negate(PointA);
            DbVector3 TriangleNormalABC = Cross(EdgeAB, EdgeAC);

            float DotAB_AO = Dot(EdgeAB, VectorToOriginFromA);
            float DotAC_AO = Dot(EdgeAC, VectorToOriginFromA);

            if (DotAB_AO <= 0f && DotAC_AO <= 0f)
            {
                GjkVertex VertexA = Simplex[2];
                Simplex.Clear();
                Simplex.Add(VertexA);
                SearchDirection = VectorToOriginFromA;
                return false;

            }

            if (DotAB_AO > 0f)
            {
                DbVector3 PerpendicularTowardAB = TripleCross(EdgeAC, EdgeAB, EdgeAB);
                float DotABRegion = Dot(PerpendicularTowardAB, VectorToOriginFromA);

                if (DotABRegion > 0f)
                {
                    Simplex.RemoveAt(0);
                    DbVector3 NewDirection = PerpendicularTowardAB;

                    if (NearZero(NewDirection))
                    {
                        NewDirection = Perp(EdgeAB);
                    }

                    SearchDirection = NewDirection;
                    return false;
                }
            }

            if (DotAC_AO > 0f)
            {
                DbVector3 PerpendicularTowardAC = TripleCross(EdgeAB, EdgeAC, EdgeAC);
                float DotACRegion = Dot(PerpendicularTowardAC, VectorToOriginFromA);

                if (DotACRegion > 0f)
                {
                    Simplex.RemoveAt(1);
                    DbVector3 NewDirection = PerpendicularTowardAC;

                    if (NearZero(NewDirection))
                    {
                        NewDirection = Perp(EdgeAC);
                    }

                    SearchDirection = NewDirection;
                    return false;
                }
            }

            DbVector3 OrientedNormal = TriangleNormalABC;

            if (Dot(OrientedNormal, VectorToOriginFromA) < 0f)
            {
                OrientedNormal = Negate(OrientedNormal);
            }

            SearchDirection = OrientedNormal;
            return false;
        }

        if (SimplexCount == 4)
        {
            DbVector3 PointA = Simplex[3].MinkowskiPoint;
            DbVector3 PointB = Simplex[2].MinkowskiPoint;
            DbVector3 PointC = Simplex[1].MinkowskiPoint;
            DbVector3 PointD = Simplex[0].MinkowskiPoint;

            DbVector3 VectorToOriginFromA = Negate(PointA);
            DbVector3 EdgeAB = Sub(PointB, PointA);
            DbVector3 EdgeAC = Sub(PointC, PointA);
            DbVector3 EdgeAD = Sub(PointD, PointA);

            DbVector3 FaceNormalABC = Cross(EdgeAB, EdgeAC);
            DbVector3 FaceNormalACD = Cross(EdgeAC, EdgeAD);
            DbVector3 FaceNormalADB = Cross(EdgeAD, EdgeAB);

            float DotABC = Dot(FaceNormalABC, VectorToOriginFromA);
            float DotACD = Dot(FaceNormalACD, VectorToOriginFromA);
            float DotADB = Dot(FaceNormalADB, VectorToOriginFromA);

            if (DotABC > 0f)
            {
                Simplex.RemoveAt(0);
                SearchDirection = FaceNormalABC;
                return false;
            }

            if (DotACD > 0f)
            {
                Simplex.RemoveAt(2);
                SearchDirection = FaceNormalACD;
                return false;
            }

            if (DotADB > 0f)
            {
                Simplex.RemoveAt(1);
                SearchDirection = FaceNormalADB;
                return false;
            }

            return true;
        }

        DbVector3 PointSingle = Simplex[0].MinkowskiPoint;
        SearchDirection = Negate(PointSingle);
        return false;
    }

    static float ComputePenetrationDepthApprox(List<ConvexHullCollider> ColliderA, DbVector3 PositionA, float YawRadiansA, List<ConvexHullCollider> ColliderB, DbVector3 PositionB, float YawRadiansB, DbVector3 Normal)
    {
        // A uses -Normal (Backwards), B uses +Normal (Forwards)
        DbVector3 SupportA = SupportWorldComplex(ColliderA, PositionA, YawRadiansA, Negate(Normal));
        DbVector3 SupportB = SupportWorldComplex(ColliderB, PositionB, YawRadiansB, Normal);

        float DistanceA = Dot(SupportA, Normal);
        float DistanceB = Dot(SupportB, Normal);

        float Gap = DistanceB - DistanceA;

        // FIX: If Gap is positive, we are overlapping. Return the Gap.
        // If Gap is negative, we are separated. Return 0.
        if (Gap <= 0f) return 0f;
        return Gap;
    }
}

File: server-rust/spacetimedb/src/Collisions/GJK/Methods.rs
use crate::*;

pub fn solve_gjk(collider_a: &Vec<ConvexHullCollider>, position_a: DbVector3, yaw_radians_a: f32, collider_b: &Vec<ConvexHullCollider>, position_b: DbVector3, yaw_radians_b: f32, result_out: &mut GjkResult, max_iterations: i32) -> bool { // Returns whether two objects are colliding with information to rebuild accurate normal and depth
    let mut simplex: Vec<GjkVertex> = Vec::with_capacity(4);
    let mut search_direction = DbVector3 { x: 0.0, y: 0.0, z: 1.0 };

    let initial_vertex = support_pair_world(collider_a, position_a, yaw_radians_a, collider_b, position_b, yaw_radians_b, search_direction);
    let initial_dot: f32 = dot(initial_vertex.minkowski_point, search_direction);

    if initial_dot <= 0.0 {
        *result_out = GjkResult { intersects: false, simplex, last_direction: negate(search_direction) };
        return false;
    }

    simplex.push(initial_vertex);
    search_direction = negate(simplex[0].minkowski_point);

    for _iteration_index in 0..max_iterations {
        let support_vertex = support_pair_world(collider_a, position_a, yaw_radians_a, collider_b, position_b, yaw_radians_b, search_direction);
        let support_dot: f32 = dot(support_vertex.minkowski_point, search_direction);

        if support_dot <= 0.0 {
            *result_out = GjkResult { intersects: false, simplex, last_direction: negate(search_direction) };
            return false;
        }

        simplex.push(support_vertex);

        if update_simplex(&mut simplex, &mut search_direction) {
            *result_out = GjkResult { intersects: true, simplex, last_direction: negate(search_direction) };
            return true;
        }
    }

    *result_out = GjkResult { intersects: false, simplex, last_direction: negate(search_direction) };
    false
}

pub fn update_simplex(simplex: &mut Vec<GjkVertex>, search_direction: &mut DbVector3) -> bool {
    let simplex_count: usize = simplex.len();

    if simplex_count == 2 {
        let point_a = simplex[1].minkowski_point;
        let point_b = simplex[0].minkowski_point;

        let segment_ba = sub(point_b, point_a);
        let vector_to_origin_from_a = negate(point_a);

        if dot(segment_ba, vector_to_origin_from_a) <= 0.0 {
            simplex[0] = simplex[1];
            simplex.pop();
            *search_direction = vector_to_origin_from_a;
            return false;
        }

        let mut new_direction = triple_cross(segment_ba, vector_to_origin_from_a, segment_ba);
        if near_zero(new_direction) { new_direction = perpendicular(segment_ba); }

        *search_direction = new_direction;
        return false;
    }

    if simplex_count == 3 {
        let point_a = simplex[2].minkowski_point;
        let point_b = simplex[1].minkowski_point;
        let point_c = simplex[0].minkowski_point;

        let edge_ab = sub(point_b, point_a);
        let edge_ac = sub(point_c, point_a);
        let vector_to_origin_from_a = negate(point_a);
        let triangle_normal_abc = cross(edge_ab, edge_ac);

        let dot_ab_ao: f32 = dot(edge_ab, vector_to_origin_from_a);
        let dot_ac_ao: f32 = dot(edge_ac, vector_to_origin_from_a);

        if dot_ab_ao <= 0.0 && dot_ac_ao <= 0.0 {
            let vertex_a = simplex[2];
            simplex.clear();
            simplex.push(vertex_a);
            *search_direction = vector_to_origin_from_a;
            return false;
        }

        if dot_ab_ao > 0.0 {
            let mut perpendicular_toward_ab = triple_cross(edge_ac, edge_ab, edge_ab);
            if dot(perpendicular_toward_ab, vector_to_origin_from_a) > 0.0 {
                simplex[0] = simplex[1];
                simplex[1] = simplex[2];
                simplex.pop();
                if near_zero(perpendicular_toward_ab) { perpendicular_toward_ab = perpendicular(edge_ab); }
                *search_direction = perpendicular_toward_ab;
                return false;
            }
        }

        if dot_ac_ao > 0.0 {
            let mut perpendicular_toward_ac = triple_cross(edge_ab, edge_ac, edge_ac);
            if dot(perpendicular_toward_ac, vector_to_origin_from_a) > 0.0 {
                simplex[1] = simplex[2];
                simplex.pop();
                if near_zero(perpendicular_toward_ac) { perpendicular_toward_ac = perpendicular(edge_ac); }
                *search_direction = perpendicular_toward_ac;
                return false;
            }
        }

        let mut oriented_normal = triangle_normal_abc;
        if dot(oriented_normal, vector_to_origin_from_a) < 0.0 { oriented_normal = negate(oriented_normal); }

        *search_direction = oriented_normal;
        return false;
    }

    if simplex_count == 4 {
        let point_a = simplex[3].minkowski_point;
        let point_b = simplex[2].minkowski_point;
        let point_c = simplex[1].minkowski_point;
        let point_d = simplex[0].minkowski_point;

        let vector_to_origin_from_a = negate(point_a);
        let edge_ab = sub(point_b, point_a);
        let edge_ac = sub(point_c, point_a);
        let edge_ad = sub(point_d, point_a);

        let face_normal_abc = cross(edge_ab, edge_ac);
        let face_normal_acd = cross(edge_ac, edge_ad);
        let face_normal_adb = cross(edge_ad, edge_ab);

        if dot(face_normal_abc, vector_to_origin_from_a) > 0.0 {
            simplex[0] = simplex[1];
            simplex[1] = simplex[2];
            simplex[2] = simplex[3];
            simplex.pop();
            *search_direction = face_normal_abc;
            return false;
        }

        if dot(face_normal_acd, vector_to_origin_from_a) > 0.0 {
            simplex[2] = simplex[3];
            simplex.pop();
            *search_direction = face_normal_acd;
            return false;
        }

        if dot(face_normal_adb, vector_to_origin_from_a) > 0.0 {
            simplex[1] = simplex[2];
            simplex[2] = simplex[3];
            simplex.pop();
            *search_direction = face_normal_adb;
            return false;
        }

        return true;
    }

    let point_single = simplex[0].minkowski_point;
    *search_direction = negate(point_single);
    false
}

pub fn support_pair_world(complex_collider_a: &Vec<ConvexHullCollider>, position_a: DbVector3, yaw_radians_a: f32, complex_collider_b: &Vec<ConvexHullCollider>, position_b: DbVector3, yaw_radians_b: f32, direction_world: DbVector3) -> GjkVertex {
    let direction_len_sq: f32 = dot(direction_world, direction_world);
    let direction_inv_len: f32 = if direction_len_sq > 1e-8 { 1.0 / direction_len_sq.sqrt() } else { 0.0 };

    let support_point_a_world = support_world_complex(complex_collider_a, position_a, yaw_radians_a, direction_world, direction_inv_len);
    let support_point_b_world = support_world_complex(complex_collider_b, position_b, yaw_radians_b, negate(direction_world), direction_inv_len);

    GjkVertex { support_point_a: support_point_a_world, support_point_b: support_point_b_world, minkowski_point: sub(support_point_a_world, support_point_b_world) }
}

pub fn support_world_complex(complex_collider: &Vec<ConvexHullCollider>, world_position: DbVector3, yaw_radians: f32, direction_world: DbVector3, direction_inv_len: f32) -> DbVector3 {
    let direction_local = rotate_around_y_axis(direction_world, -yaw_radians);
    let support_local_point = support_local_complex(complex_collider, direction_local, direction_inv_len);
    let support_world_rotated = rotate_around_y_axis(support_local_point, yaw_radians);
    add(support_world_rotated, world_position)
}

pub fn support_local_complex(complex_collider: &Vec<ConvexHullCollider>, direction_local: DbVector3, direction_inv_len: f32) -> DbVector3 {
    let mut best_dot: f32 = f32::NEG_INFINITY;
    let mut best_point = DbVector3 { x: 0.0, y: 0.0, z: 0.0 };

    for index in 0..complex_collider.len() {
        let (hull_support_point, hull_support_dot) = support_local_and_dot(&complex_collider[index], direction_local, direction_inv_len);

        if hull_support_dot > best_dot {
            best_dot = hull_support_dot;
            best_point = hull_support_point;
        }
    }

    best_point
}

pub fn support_local_and_dot(collider: &ConvexHullCollider, direction: DbVector3, direction_inv_len: f32) -> (DbVector3, f32) {
    let vertices: &Vec<DbVector3> = &collider.vertices_local;

    let mut best_vertex_index: usize = 0;
    let mut best_dot_product: f32 = dot(vertices[0], direction);

    for vertex_index in 1..vertices.len() {
        let dot_product: f32 = dot(vertices[vertex_index], direction);
        if dot_product > best_dot_product {
            best_dot_product = dot_product;
            best_vertex_index = vertex_index;
        }
    }

    let mut best_vertex = vertices[best_vertex_index];

    let margin: f32 = collider.margin;
    if margin > 0.0 && direction_inv_len > 0.0 {
        best_vertex = DbVector3 { x: best_vertex.x + direction.x * direction_inv_len * margin, y: best_vertex.y + direction.y * direction_inv_len * margin, z: best_vertex.z + direction.z * direction_inv_len * margin };
        best_dot_product += margin / direction_inv_len;
    }

    (best_vertex, best_dot_product)
}
